src/IECoreScene/SceneAlgo.cpp

```diff
@@ -0,0 +1,285 @@
+		}
+	}
+
+	if( flags & SceneAlgo::Sets && isRoot )
+	{
+		SceneInterface::NameList setNames = src->setNames();
+		copyInfo.setCount += setNames.size();
+		for( const auto &setName : setNames )
+		{
+			PathMatcher set = src->readSet( setName );
+			if( dst )
+			{
+				dst->writeSet( setName, set );
+			}
+		}
+	}
+
+	if( flags & SceneAlgo::Objects && src->hasObject() )
+	{
+		IECore::ConstObjectPtr obj = src->readObject( time );
+
+		if( IECoreScene::MeshPrimitive::ConstPtr mesh = IECore::runTimeCast<const IECoreScene::MeshPrimitive>( obj ) )
+		{
+			copyInfo.polygonCount += mesh->numFaces();
+		}
+		else if( IECoreScene::CurvesPrimitive::ConstPtr curves = IECore::runTimeCast<const IECoreScene::CurvesPrimitive>( obj ) )
+		{
+			copyInfo.curveCount += curves->numCurves();
+		}
+		else if( IECoreScene::PointsPrimitive::ConstPtr points = IECore::runTimeCast<const IECoreScene::PointsPrimitive>( obj ) )
+		{
+			copyInfo.pointCount += points->getNumPoints();
+		}
+		if( dst )
+		{
+			dst->writeObject( obj.get(), time );
+		}
+	}
+
+	return copyInfo;
+
+}
+
+template<typename LocationFn>
+void copy( ConstSceneInterfacePtr src, SceneInterfacePtr dst, double time, unsigned int flags, LocationFn &locationFn )
+{
+	locationFn( src, dst, time, flags );
+
+	SceneInterface::NameList childNames;
+	src->childNames( childNames );
+
+	for( const auto &childName : childNames )
+	{
+		SceneInterfacePtr dstChild = dst ? dst->child( childName, SceneInterface::CreateIfMissing ) : nullptr;
+		::copy( src->child( childName ), dstChild, time, flags, locationFn );
+	}
+}
+
+template<typename LocationFn>
+void copy( ConstSceneInterfacePtr src, SceneInterfacePtr dst, double time, int threadCount, unsigned int flags, LocationFn &locationFn )
+{
+
+	if( threadCount > 1 && !dst )
+	{
+
+		tbb::task_scheduler_init taskSchedulerInit( threadCount );
+		Task<LocationFn> *task = new( tbb::task::allocate_root() ) Task<LocationFn>( src, dst, locationFn, time, flags );
+		tbb::task::spawn_root_and_wait( *task );
+	}
+	else
+	{
+		copy( src, dst, time, flags, locationFn );
+	}
+}
+
+} // namespace
+
+namespace IECoreScene
+{
+namespace SceneAlgo
+{
+
+CopyStats copy( ConstSceneInterfacePtr src, SceneInterfacePtr dst, int startFrame, int endFrame, float frameRate, int threadCount, unsigned int flags )
+{
+	std::atomic<size_t> locationCount (0);
+	::CopyInfo<std::atomic<size_t> > copyInfos;
+
+	auto locationFn = [&locationCount, &copyInfos] ( ConstSceneInterfacePtr src, SceneInterfacePtr dst, double time, unsigned int flags )
+	{
+		locationCount++;
+		::CopyInfo<size_t> copyInfo = ::copyLocation( src, dst, time, flags);
+
+		copyInfos.polygonCount += copyInfo.polygonCount;
+		copyInfos.tagCount += copyInfo.tagCount;
+		copyInfos.attributeCount += copyInfo.attributeCount;
+		copyInfos.curveCount += copyInfo.curveCount;
+		copyInfos.pointCount += copyInfo.pointCount;
+	};
+
+	for( int f = startFrame; f <= endFrame; ++f )
+	{
+		double time = f / frameRate;
+		if (f == startFrame)
```

Comment:
As an optimisation, could we remove Tags from the flags for all but the first frame, since tags are constant across time?